File: apps/aurix-core-services/src/services/resilience-engine/AllocationOptimizer.ts
/**
 * AurixResilienceEngine — AllocationOptimizer
 *
 * Computes safe allocation weights for a user's portfolio to maximize
 * financial resilience. The weighting logic is inspired by portfolio
 * optimization theory: assets are scored by their contribution to
 * stability, then normalized to a 10,000 basis-point weight sum so the
 * output can be consumed directly by on-chain rebalancing logic.
 *
 * Design pattern: three-phase pipeline
 *   1. Collect asset price/volatility data
 *   2. Score each asset for resilience contribution
 *   3. Normalize weights and return allocation recommendations
 *
 * This is optimization for SAFETY, not yield.
 * Stablecoins (QUSDC) are systematically over-weighted.
 * Volatile assets are systematically under-weighted.
 */

// ── Types ─────────────────────────────────────────────────────────────────────

export interface AssetDataPoint {
  tokenId:    string;
  symbol:     string;
  valueUsd:   number;
  volatility: number;   // 30-day price volatility, 0.0 – 1.0
  isStable:   boolean;  // true for QUSDC and other stablecoins
  chainId:    number;
}

export interface AllocationWeight {
  tokenId:       string;
  symbol:        string;
  weightBps:     number;  // basis points: sum must equal 10_000
  safetyScore:   number;  // 0–100 contribution to resilience
  reasoning:     string;
}

export interface AllocationResult {
  weights:              AllocationWeight[];
  stableAllocationPct:  number;   // % in stablecoins
  diversificationScore: number;   // 0–100
  recommendedAction:    string;
  computedAt:           number;   // unix timestamp
}

// ── Safety scoring coefficients ───────────────────────────────────────────────

const STABLE_BASE_WEIGHT   = 4_000;  // stablecoins start at 40% of total
const VOLATILE_PENALTY     = 0.65;   // volatile asset weight multiplied by this
const MIN_STABLE_PCT       = 0.15;   // force at least 15% in stablecoins
const MULTI_CHAIN_BONUS    = 0.05;   // 5% resilience bonus per additional chain

// ── AllocationOptimizer ───────────────────────────────────────────────────────

export class AllocationOptimizer {

  /**
   * Compute safe allocation weights from a list of asset data points.
   *
   * Algorithm (three-phase, adapted from portfolio optimization patterns):
   *   Phase 1: Raw score each asset (stability, value, volatility)
   *   Phase 2: Compute proportional weights summing to 10,000 bps
   *   Phase 3: Enforce minimum stable floor, normalize again
   */
  static optimize(assets: AssetDataPoint[]): AllocationResult {
    if (assets.length === 0) {
      return {
        weights:              [],
        stableAllocationPct:  0,
        diversificationScore: 0,
        recommendedAction:    "No assets detected. Connect QIE Pass to begin.",
        computedAt:           Math.floor(Date.now() / 1000),
      };
    }

    const totalValue = assets.reduce((sum, a) => sum + a.valueUsd, 0);
    if (totalValue === 0) {
      return this._zeroValueResult(assets);
    }

    // Phase 1: Score each asset
    const rawScores = assets.map(a => ({
      ...a,
      rawScore: this._scoreAsset(a, totalValue),
    }));

    const totalRaw = rawScores.reduce((s, a) => s + a.rawScore, 0);

    // Phase 2: Proportional weights (10,000 bps)
    let weights: AllocationWeight[] = rawScores.map(a => ({
      tokenId:     a.tokenId,
      symbol:      a.symbol,
      weightBps:   Math.round((a.rawScore / totalRaw) * 10_000),
      safetyScore: Math.round(Math.min(a.rawScore / totalRaw * 100, 100)),
      reasoning:   this._explain(a),
    }));

    // Phase 3: Enforce minimum stable floor
    weights = this._enforceStableFloor(weights, assets);

    // Normalize to exactly 10,000
    weights = this._normalize(weights);

    const stableWeight = weights
      .filter(w => assets.find(a => a.tokenId === w.tokenId)?.isStable)
      .reduce((s, w) => s + w.weightBps, 0);

    const stableAllocationPct = stableWeight / 100;

    return {
      weights,
      stableAllocationPct,
      diversificationScore: this._diversificationScore(assets, weights),
      recommendedAction:    this._recommendAction(stableAllocationPct),
      computedAt:           Math.floor(Date.now() / 1000),
    };
  }

  // ── Private helpers ──────────────────────────────────────────────────────────

  private static _scoreAsset(a: AssetDataPoint, totalValue: number): number {
    const valueShare   = a.valueUsd / totalValue;
    const stabilityMod = a.isStable ? 1.5 : (1 - a.volatility * VOLATILE_PENALTY);
    return Math.max(valueShare * stabilityMod * 100, 0.01);
  }

  private static _enforceStableFloor(
    weights: AllocationWeight[],
    assets:  AssetDataPoint[],
  ): AllocationWeight[] {
    const stableIds   = new Set(assets.filter(a => a.isStable).map(a => a.tokenId));
    const stableBps   = weights.filter(w => stableIds.has(w.tokenId)).reduce((s, w) => s + w.weightBps, 0);
    const floorBps    = Math.round(MIN_STABLE_PCT * 10_000);

    if (stableBps >= floorBps || stableIds.size === 0) return weights;

    const deficit     = floorBps - stableBps;
    const nonStable   = weights.filter(w => !stableIds.has(w.tokenId));
    const totalNonBps = nonStable.reduce((s, w) => s + w.weightBps, 0);

    return weights.map(w => {
      if (stableIds.has(w.tokenId)) {
        return { ...w, weightBps: w.weightBps + Math.round(deficit / stableIds.size) };
      }
      const reduction = Math.round((w.weightBps / (totalNonBps || 1)) * deficit);
      return { ...w, weightBps: Math.max(w.weightBps - reduction, 0) };
    });
  }

  private static _normalize(weights: AllocationWeight[]): AllocationWeight[] {
    const total = weights.reduce((s, w) => s + w.weightBps, 0);
    if (total === 0 || total === 10_000) return weights;
    const diff  = 10_000 - total;
    // Apply diff to largest weight
    const maxIdx = weights.reduce((mi, w, i) => w.weightBps > weights[mi].weightBps ? i : mi, 0);
    return weights.map((w, i) => i === maxIdx ? { ...w, weightBps: w.weightBps + diff } : w);
  }

  private static _diversificationScore(assets: AssetDataPoint[], weights: AllocationWeight[]): number {
    const chains   = new Set(assets.map(a => a.chainId)).size;
    const herfindahl = weights.reduce((s, w) => s + Math.pow(w.weightBps / 10_000, 2), 0);
    // Herfindahl: 1 = fully concentrated, 0 = perfectly diversified
    const baseScore = Math.round((1 - herfindahl) * 80);
    const chainBonus = Math.round(Math.min((chains - 1) * MULTI_CHAIN_BONUS * 100, 20));
    return Math.min(baseScore + chainBonus, 100);
  }

  private static _explain(a: AssetDataPoint & { rawScore: number }): string {
    if (a.isStable)            return `${a.symbol} is a stablecoin — weighted up for resilience`;
    if (a.volatility > 0.6)   return `${a.symbol} has high volatility (${(a.volatility * 100).toFixed(0)}%) — weighted down`;
    if (a.volatility < 0.2)   return `${a.symbol} has low volatility — good resilience contribution`;
    return `${a.symbol} has moderate volatility — balanced allocation`;
  }

  private static _recommendAction(stablePct: number): string {
    if (stablePct < 10) return "CRITICAL: Swap assets to QUSDC immediately — stablecoin buffer dangerously low";
    if (stablePct < 20) return "Increase QUSDC allocation to at least 20% for adequate protection";
    if (stablePct < 40) return "Stablecoin buffer is adequate. Consider adding cross-chain diversification.";
    return "Allocation is well-balanced for financial resilience.";
  }

  private static _zeroValueResult(assets: AssetDataPoint[]): AllocationResult {
    return {
      weights: assets.map(a => ({ tokenId: a.tokenId, symbol: a.symbol, weightBps: 0, safetyScore: 0, reasoning: "Zero portfolio value" })),
      stableAllocationPct:  0,
      diversificationScore: 0,
      recommendedAction:    "Portfolio value is zero. Fund your wallet to begin resilience optimization.",
      computedAt:           Math.floor(Date.now() / 1000),
    };
  }
}

File: apps/aurix-oracle/engines/allocation_optimizer.py
"""
AurixResilienceEngine — AllocationOptimizer (Python)

Computes safe portfolio allocation weights for Aurix users.

Engineering pattern:
  - Three-phase pipeline: price fetch → optimization → result
  - Weight normalization to 10,000 basis points
  - Classical allocation logic (classical math only)

Aurix implementation:
  - Optimizes for SAFETY, not yield or profit
  - Stablecoins (QUSDC) are systematically over-weighted
  - Volatile assets are penalized proportional to volatility
  - Output is a ranked allocation recommendation for the user
"""

from typing import List, Dict, Any
import math


# ── Safety scoring constants ──────────────────────────────────────────────────

STABLE_WEIGHT_BOOST    = 1.5    # stablecoins get 1.5x their value share
VOLATILE_PENALTY       = 0.65   # volatile assets: weight * this factor
MIN_STABLE_BPS         = 1500   # floor: at least 15% in stablecoins
TOTAL_BPS              = 10_000


# ── AllocationOptimizer ───────────────────────────────────────────────────────

class AllocationOptimizer:
    """
    Three-phase allocation optimizer for financial resilience.

    Input:  List of asset dicts with tokenId, symbol, valueUsd,
            volatility (0.0–1.0), isStable (bool), chainId
    Output: Dict with weights (list of {tokenId, symbol, weightBps}),
            stableAllocationPct, diversificationScore, recommendedAction
    """

    @staticmethod
    def optimize(assets: List[Dict[str, Any]]) -> Dict[str, Any]:
        if not assets:
            return {
                "weights": [],
                "stableAllocationPct": 0,
                "diversificationScore": 0,
                "recommendedAction": "No assets detected. Connect QIE Pass to begin.",
                "computedAt": AllocationOptimizer._now()
            }

        total_value = sum(a.get("valueUsd", 0) for a in assets)
        if total_value <= 0:
            return AllocationOptimizer._zero_result(assets)

        # Phase 1: Score each asset
        raw_scores = [
            AllocationOptimizer._score_asset(a, total_value)
            for a in assets
        ]
        total_raw = sum(raw_scores) or 1.0

        # Phase 2: Proportional weights (10,000 bps)
        weights = []
        for i, asset in enumerate(assets):
            bps = round((raw_scores[i] / total_raw) * TOTAL_BPS)
            weights.append({
                "tokenId":   asset.get("tokenId", ""),
                "symbol":    asset.get("symbol", ""),
                "weightBps": bps,
                "safetyScore": round(min((raw_scores[i] / total_raw) * 100, 100)),
                "reasoning": AllocationOptimizer._explain(asset),
            })

        # Phase 3: Enforce minimum stable floor
        weights = AllocationOptimizer._enforce_stable_floor(weights, assets)

        # Normalize to exactly 10,000 bps
        weights = AllocationOptimizer._normalize(weights)

        stable_bps = sum(
            w["weightBps"] for w, a in zip(weights, assets) if a.get("isStable")
        )
        stable_pct = stable_bps / 100.0

        return {
            "weights": weights,
            "stableAllocationPct": stable_pct,
            "diversificationScore": AllocationOptimizer._diversification_score(assets, weights),
            "recommendedAction": AllocationOptimizer._recommend(stable_pct),
            "computedAt": AllocationOptimizer._now(),
        }

    # ── Private helpers ───────────────────────────────────────────────────────

    @staticmethod
    def _score_asset(asset: Dict[str, Any], total_value: float) -> float:
        value_share  = asset.get("valueUsd", 0) / total_value
        is_stable    = asset.get("isStable", False)
        volatility   = asset.get("volatility", 0.5)
        stability    = STABLE_WEIGHT_BOOST if is_stable else (1 - volatility * VOLATILE_PENALTY)
        return max(value_share * stability * 100, 0.01)

    @staticmethod
    def _enforce_stable_floor(
        weights: List[Dict[str, Any]],
        assets:  List[Dict[str, Any]],
    ) -> List[Dict[str, Any]]:
        stable_ids  = {a.get("tokenId") for a in assets if a.get("isStable")}
        stable_bps  = sum(w["weightBps"] for w in weights if w["tokenId"] in stable_ids)
        deficit     = MIN_STABLE_BPS - stable_bps
        if deficit <= 0 or not stable_ids:
            return weights

        non_stable  = [w for w in weights if w["tokenId"] not in stable_ids]
        total_ns    = sum(w["weightBps"] for w in non_stable) or 1

        result = []
        for w in weights:
            if w["tokenId"] in stable_ids:
                result.append({**w, "weightBps": w["weightBps"] + deficit // len(stable_ids)})
            else:
                reduction = round((w["weightBps"] / total_ns) * deficit)
                result.append({**w, "weightBps": max(w["weightBps"] - reduction, 0)})
        return result

    @staticmethod
    def _normalize(weights: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        total = sum(w["weightBps"] for w in weights)
        if total == TOTAL_BPS or total == 0:
            return weights
        diff     = TOTAL_BPS - total
        max_idx  = max(range(len(weights)), key=lambda i: weights[i]["weightBps"])
        result   = list(weights)
        result[max_idx] = {**result[max_idx], "weightBps": result[max_idx]["weightBps"] + diff}
        return result

    @staticmethod
    def _diversification_score(
        assets:  List[Dict[str, Any]],
        weights: List[Dict[str, Any]],
    ) -> int:
        chains     = len({a.get("chainId", 1) for a in assets})
        hhi        = sum(math.pow(w["weightBps"] / TOTAL_BPS, 2) for w in weights)
        base_score = round((1 - hhi) * 80)
        chain_bonus = min((chains - 1) * 5, 20)
        return min(base_score + chain_bonus, 100)

    @staticmethod
    def _recommend(stable_pct: float) -> str:
        if stable_pct < 10:
            return "CRITICAL: Swap assets to QUSDC immediately — stablecoin buffer dangerously low"
        if stable_pct < 20:
            return "Increase QUSDC allocation to at least 20% for adequate protection"
        if stable_pct < 40:
            return "Stablecoin buffer is adequate. Consider cross-chain diversification."
        return "Allocation is well-balanced for financial resilience."

    @staticmethod
    def _explain(asset: Dict[str, Any]) -> str:
        if asset.get("isStable"):
            return f"{asset.get('symbol')} is a stablecoin — weighted up for resilience"
        vol = asset.get("volatility", 0.5)
        if vol > 0.6:
            return f"{asset.get('symbol')} has high volatility ({vol*100:.0f}%) — weighted down"
        return f"{asset.get('symbol')} has moderate volatility — balanced allocation"

    @staticmethod
    def _zero_result(assets: List[Dict[str, Any]]) -> Dict[str, Any]:
        return {
            "weights": [{"tokenId": a.get("tokenId"), "symbol": a.get("symbol"), "weightBps": 0} for a in assets],
            "stableAllocationPct": 0,
            "diversificationScore": 0,
            "recommendedAction": "Portfolio value is zero. Fund your wallet to begin.",
            "computedAt": AllocationOptimizer._now(),
        }

    @staticmethod
    def _now() -> int:
        import time
        return int(time.time())
